### PR TITLE
Add Parallels Access Latest

### PR DESCRIPTION
**Pull request by samshadwell**
Does not currently work. I think this is related to me not properly running the installer script

### CODE REVIEW

**Comment by vitorgalvao** [comment]
First things first. Why do you have both an `app` and an `installer`?

**Comment by vitorgalvao** [comment]
1. Why do you have both an `app` and an `installer`?
2. What do you mean by “doesn’t work”? Doesn’t pass `audit`, doesn’t install? What?
3. What are the exact errors you are getting?

**Comment by samshadwell** [comment]
1. So initially I just had the `installer` but the installation process of this particular application seems kind of strange. The script that I reference doesn't actually install the application itself it seems, but rather installs some libraries or something (I apologize for the vague language, it's not clear to me what exactly is going on). I added the `app` bit because the script didn't result in anything being installed to /Applications or being symlinked there.

2. Doesn't work as in, in its current form it passes audit, and appears to install. But starting the application doesn't have the intended effect. It spins up a couple of application processes, but they don't appear to do anything. Definitely not the same effect as when I installed it traditionally.

3. I'm away from my Mac right now, but I'll update with those shortly.

**Comment by vitorgalvao** [comment]
Walk me through the traditional installation process, then. How do you install it?

**Comment by samshadwell** [comment]
Switched back to my mac, the installation process currently has the following output:

> ==> Downloading https://www.parallels.com/directdownload/pax-agent/mac/
######################################################################## 100.0%
==> No checksum defined for Cask parallels-access, skipping verification
==> Running installer script Install.app/Contents/MacOS/PMInstall
Password:
==> 2016-03-21 15:21:49.135 PMInstall[9099:307910] Failed to connect (colorGridView) outlet from (NSApplication) to (NSColorPickerGridView): missing setter or instance variable
==> 2016-03-21 15:21:49.135 PMInstall[9099:307910] Failed to connect (view) outlet from (NSApplication) to (NSColorPickerGridView): missing setter or instance variable
==> objc[9101]: Class WebCoreMovieObserver is implemented in both /System/Library/Frameworks/WebKit.framework/Versions/A/Frameworks/WebCore.framework/Versions/A/WebCore and /opt/homebrew-cask/Caskroom/parallels-access/latest/Parallels Acc
==> objc[9101]: Class WebCoreSharedBufferData is implemented in both /System/Library/Frameworks/WebKit.framework/Versions/A/Frameworks/WebCore.framework/Versions/A/WebCore and /opt/homebrew-cask/Caskroom/parallels-access/latest/Parallels
==> objc[9101]: Class WebVideoFullscreenWindow is implemented in both /System/Library/Frameworks/WebKit.framework/Versions/A/Frameworks/WebCore.framework/Versions/A/WebCore and /opt/homebrew-cask/Caskroom/parallels-access/latest/Parallels
==> objc[9101]: Class WebVideoFullscreenController is implemented in both /System/Library/Frameworks/WebKit.framework/Versions/A/Frameworks/WebCore.framework/Versions/A/WebCore and /opt/homebrew-cask/Caskroom/parallels-access/latest/Paral
==> objc[9101]: Class WebVideoFullscreenHUDWindowController is implemented in both /System/Library/Frameworks/WebKit.framework/Versions/A/Frameworks/WebCore.framework/Versions/A/WebCore and /opt/homebrew-cask/Caskroom/parallels-access/lat
==> objc[9101]: Class WebVideoFullscreenHUDWindow is implemented in both /System/Library/Frameworks/WebKit.framework/Versions/A/Frameworks/WebCore.framework/Versions/A/WebCore and /opt/homebrew-cask/Caskroom/parallels-access/latest/Parall
==> objc[9101]: Class WebWindowFadeAnimation is implemented in both /System/Library/Frameworks/WebKit.framework/Versions/A/Frameworks/WebCore.framework/Versions/A/WebCore and /opt/homebrew-cask/Caskroom/parallels-access/latest/Parallels A
==> objc[9101]: Class WebWindowScaleAnimation is implemented in both /System/Library/Frameworks/WebKit.framework/Versions/A/Frameworks/WebCore.framework/Versions/A/WebCore and /opt/homebrew-cask/Caskroom/parallels-access/latest/Parallels
==> Mar 21 15:21:49  /opt/homebrew-cask/Caskroom/parallels-access/latest/Parallels Access.app/Contents/MacOS/pm_ctl[9109] <Error>: Checking
==> Mar 21 15:21:49  /opt/homebrew-cask/Caskroom/parallels-access/latest/Parallels Access.app/Contents/MacOS/pm_ctl[9128] <Error>: '/opt/homebrew-cask/Caskroom/parallels-access/latest/Parallels Access.app/Contents/MacOS/Parallels Service'
==> 2016-03-21 15:22:22.822 prl_deskctl_wizard[9101:307923] modalSession has been exited prematurely - check for a reentrant call to endModalSession:
==> 2016-03-21 15:22:22.929 prl_deskctl_wizard[9101:307923] modalSession has been exited prematurely - check for a reentrant call to endModalSession:
==> Symlinking App 'Parallels Access.app' to '/Applications/Parallels Access.app'
🍺  parallels-access staged at '/opt/homebrew-cask/Caskroom/parallels-access/latest' (217 files, 260M)

After the Password prompt, a license agreement windows opens (which I have been accepting).

**Comment by samshadwell** [comment]
I'll go through normal installation in a second, and respond here with that.

**Comment by samshadwell** [comment]
Normal install process:
1. Mount the .dmg, it opens with a large button to run Install.app
2. It runs and prompts me to accept the license
3. After accepting it requests access "to perform one-time initialization"
4. Window displays that it is "initializing"
5. After this, the application opens. It has a persistent icon in the title bar although it doesn't have anything in the dock.

**Comment by samshadwell** [comment]
I should mention, it also appears to install "Parallels Access.app" to `/Applications`

File: Casks/parallels-access.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+cask 'parallels-access' do
+  version :latest
+  sha256 :no_check
+
+  url 'https://www.parallels.com/directdownload/pax-agent/mac/'
```
Comment (vitorgalvao) [inline Casks/parallels-access.rb]: `url "http://download.parallels.com/pmobile/v#{version.major}/#{version.major_minor_patch}/ParallelsAccess-#{version}-mac.dmg"`.

File: Casks/parallels-access.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+cask 'parallels-access' do
+  version :latest
```
Comment (vitorgalvao) [inline Casks/parallels-access.rb]: `version '3.0.2-30719'`.

File: Casks/parallels-access.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+  sha256 :no_check
```
Comment (vitorgalvao) [inline Casks/parallels-access.rb]: `sha256` also needs to be set. You can find what it is by running `shasum -a 256` on the downloaded file.

File: Casks/parallels-access.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+  url 'https://www.parallels.com/directdownload/pax-agent/mac/'
+  name 'Parallels Access'
+  homepage 'http://www.parallels.com/products/access/'
+  license :closed
+
+  app 'Parallels Access.app'
```
Comment (vitorgalvao) [inline Casks/parallels-access.rb]: This needs to go. There’s no such app inside the dmg.

File: Casks/parallels-access.rb [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+  installer script: 'Install.app/Contents/MacOS/PMInstall',
```
Comment (vitorgalvao) [inline Casks/parallels-access.rb]: This calls a GUI, so we can’t use it. Use instead [`installer manual:`](https://github.com/caskroom/homebrew-cask/blob/master/doc/cask_language_reference/stanzas/installer.md#installer-manual)

**Comment by vitorgalvao** [comment]
It’s missing `uninstall`. There are [scripts](https://github.com/caskroom/homebrew-cask/tree/master/developer/bin) and [documentation](https://github.com/caskroom/homebrew-cask/blob/master/doc/cask_language_reference/stanzas/uninstall.md) geared for that.

File: Casks/parallels-access.rb (hunk 1 above)
Comment (samshadwell) [inline Casks/parallels-access.rb]: How did you manage to find this download link?

Comment (vitorgalvao) [inline Casks/parallels-access.rb]: Your link redirects to this one. Just opened yours and then checked Chrome’s download manager.

**Comment by samshadwell** [comment]
I'm having issues trying to install this cask now, would you happen to know what could cause that?

I'm getting

> $ brew cask install parallels-access
Error: No available Cask for parallels-access
Error: nothing to install

I've already followed the "Getting Set Up to Contribute" section.

**Comment by samshadwell** [comment]
The final results is actually the same as before, where the application doesn't launch properly after installing with the new recipe.

I poked around in the .dmg, and in addition to the `Install.app` there's also `Parallels Access.app`, and I'm guessing the exclusion of this in the install process is somehow related to the installs not working. What is protocol for this type of installer? I can't seem to find any relevant examples in existing casks.

File: Casks/parallels-access.rb [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+            quit:      [
+                         'com.parallels.Access.FileSharingExtension',
+                         'com.parallels.inputmethod.ParallelsIM',
+                         'com.parallels.mobile',
+                         'com.parallels.mobile.prl_deskctl_agent',
+                         'com.parallels.pmobile.dispatcher',
+                         'com.parallels.virtualhid',
+                         'com.parallels.virtualsound',
+                       ],
+
```
Comment (vitorgalvao) [inline Casks/parallels-access.rb]: This empty line shouldn’t be here.

**Comment by vitorgalvao** [comment]
> The final results is actually the same as before, where the application doesn't launch properly after installing with the new recipe.

That’s most likely an issue with your setup, then. This new setup doesn’t actually do anything, it just tells you to manually click the installer.

**Comment by samshadwell** [comment]
Could this be an issue with the installer expecting that `Parallels Access.app` is in the same directory as it? I feel pretty confident that this is more than just a personal setup issue, if I can reliably install, uninstall, and reinstall the application normally, but not do so via my cask.

**Comment by samshadwell** [comment]
Is there any way to move the `Parallels Access.app` alongside the installer app?

**Comment by vitorgalvao** [comment]
> Is there any way to move the `Parallels Access.app` alongside the installer app?

It is moved. Everything inside the dmg is. `cd` into `/opt/homebrew-cask/Caskroom/parallels-access/3.0.2-30719` and you’ll see. I can reproduce your error, though.

Try copying it outside the dmg yourself and run it. It’ll say it can only be installed from inside the dmg. We’re not prepared for that, so we’ll have to pass on this cask, it seems. Leaving it open for a while in case other maintainer has an idea.

**Comment by vitorgalvao** [comment]
> Is there any way to move the `Parallels Access.app` alongside the installer app?

It is moved. Everything inside the dmg is. `cd` into `/opt/homebrew-cask/Caskroom/parallels-access/3.0.2-30719` and you’ll see. I can reproduce your error, though.

Try copying it outside the dmg yourself and run it. It’ll say it can only be installed from inside the dmg. We’re not prepared for that, so we’ll have to pass on this cask, it seems. Leaving it open for a while in case other @caskroom/maintainers has an idea.

File: Casks/parallels-access.rb [hunk 7]
```diff
@@ -0,0 +1,37 @@
+cask 'parallels-access' do
+  version '3.0.2-30719'
+  sha256 'afafadd6f3e384c0e421b10eb827f83fa65b60206c938431c9c20083d6a0f252'
+
+  url "http://download.parallels.com/pmobile/v#{version.major}/#{version.major_minor_patch}/ParallelsAccess-#{version}-mac.dmg"
```
Comment (victorpopkov) [inline Casks/parallels-access.rb]: The HTTPS version is available, which is more preferred.

**Comment by victorpopkov** [comment]
Actually, the `Install.app` doesn't do much but just calls the `Parallels Access.app` which is also located in the DMG image file, but it's hidden, just like in the [parallels-desktop.rb](https://github.com/caskroom/homebrew-cask/blob/master/Casks/parallels-desktop.rb). The installation procedure is done by the `Parallels Access.app` itself. But the main problem is that just symlinking the app and opening it after doesn't start the installation procedure, so it needs to be called manually.

The installation is done by the `Parallels Access.app/Contents/MacOS/pm_ctl` script when `instance_install` argument is passed. So instead of the manual installation we can use `installer script` instead:
```ruby
  installer script: 'Parallels Access.app/Contents/MacOS/pm_ctl',
            args:   %w[instance_install],
            sudo:   true
```

However, the are no uninstallation script bundled in the app. At least, I didn't found any, so it should be done manually. In my opinion, the correct uninstallation should look like this:
```ruby
  uninstall launchctl: [
                         'com.parallels.mobile.startgui.launchagent',
                         'com.parallels.mobile.startgui.launchagent',
                         'com.parallels.mobile.dispatcher.launchdaemon',
                         'com.parallels.mobile.kextloader.launchdaemon',
                         'com.parallels.mobile.prl_deskctl_agent.309792.E6A12658-F098-4D8B-957F-8997B0D58B40',
                         'com.parallels.mobile.prl_deskctl_agent.launchagent',
                       ],
            quit:      [
                         'com.parallels.inputmethod.ParallelsIM',
                       ],
            signal:    [
                         ['TERM', 'com.parallels.mobile'],
                         ['TERM', 'com.parallels.mobile.prl_deskctl_agent'],
                       ],
            kext:      [
                         'com.parallels.virtualhid',
                         'com.parallels.virtualsound',
                       ],
            delete:    '/Applications/Parallels Access.app'
```
In addition, I would also add the `zap` to the cask:
```ruby
  zap delete: [
                '~/Library/Cookies/com.parallels.mobile.prl_deskctl_agent.binarycookies',
                '~/Library/Group Containers/4C6364ACXT.com.parallels.Access',
                '~/Library/Preferences/com.parallels.Parallels Access.plist',
                '~/Library/Preferences/com.parallels.Parallels Access.plist.sdb',
                '~/Library/Preferences/com.parallels.mobile.plist',
              ]
```
I'm not sure, that this variant is 100% correct, but most likely this should do the trick. Let me know, if it works.

**Comment by vitorgalvao** [comment]
@samshadwell If you make the changes pointed out by @victorpopkov, this can be merged.

**Comment by vitorgalvao** [comment]
Thank you for the changes. Could you please [squash your commits](http://davidwalsh.name/squash-commits-git)? Please @mention me after you do, so I get an email to check back.

**Comment by samshadwell** [comment]
@vitorgalvao maybe I'm missing something, but I think that I did squash them? I followed that process at least. Also, do you know what might be going on with the Travis CI build? Looking at the log files, I can't even tell that it's attempting to do anything with my formula, and it's the only file that I have in my commits here.

**Comment by vitorgalvao** [comment]
Travis has been acting up. I’ll verify and merge locally (you’ll see a predefined message when it’s done).

**Comment by vitorgalvao** [comment]
Thank you for the contribution. It needed some fixes, so they were made in 253fc55d680a394073dcd4a98500911179d09c12. Your contribution is still included (and still credited to you), with the appropriate modifications. Please feel free to ask about any of the changes.

**Pull request closed without merging**